### ISSUE TITLE
Can you provide an example on CIFAR or any image dataset using scatnetgpu?

### ISSUE THREAD

**Issue by deepakanandece** [oinegue/scatnetgpu#2]
Hello,
Can you provide an example on CIFAR or any image dataset using scatnetgpu?

**Comment by deepakanandece** [comment]
Also the number of output channels is not given..new parameters M and L is different from pyscatwave repo?

**Comment by oinegue** [comment]
Hi, thanks for your interest in this library.
To get the scattering network representation of the CIFAR dataset, firstly you should load the dataset, [for example from keras](https://keras.io/datasets/#cifar10-small-image-classification) (any other source is ok as long as you have your images as numpy arrays).
Now say that you have the train set in `x_train` with shape (num_samples, 3, 32, 32). This library works with channels as last axis, so move them with `x_train = np.moveaxis(x_train, 1, -1)`.
Finally you can get the scatnet representation with the example code provided in the README, so the full code could be:
```python
from keras.datasets import cifar10
(x_train, y_train), (x_test, y_test) = cifar10.load_data()
x_train = np.moveaxis(x_train, 1, -1)
x_test = np.moveaxis(x_test, 1, -1)

from scatnetgpu import ScatNet
sn = ScatNet(M=2, J=4, L=6)
sn_train = []
for img in x_train:
  sn_train.append(sn.transform(img))
sn_test = []
for img in x_test:
  sn_test.append(sn.transform(img))
```
Please note that scattering networks provides only a representation of your data. A classifier (e.g. SVM) is required if you want to train a model.

For the second question, as I said on the README this library is compatible with [the original scatnet library](https://github.com/scatnet/scatnet). I've never used pyscatwave so I can't help you about the different output shape.

Anyway, you could look at [my thesis work](http://amslaurea.unibo.it/id/eprint/12261) that used this library, maybe it could guide you a little.

**Comment by deepakanandece** [comment]
Thanks for prompt reply. The transform step returns a list of list. It has several keyword as indices.  Which part is the features extracted from images?  Transform  gives a list of length one which contains lists of (3,3) which is further bifurcated. As far as i know none of the machine learning packages takes lists as input. Can you give some documentation as in which index corresponds to what?

**Comment by oinegue** [comment]
This particular output structure was chosen to keep the compatibility with the original matlab library.

First list is long 3 because data have 3 channels. Each channel is transformed independently so you have one list for each channel.
At second level, you have 3 lists because scatnet gives you M+1 representations.
For each element at this level there is a [structured array](https://docs.scipy.org/doc/numpy/user/basics.rec.html) that contains transformed signals (key `signal`) as well as path of applied filters (keys `j` and `l`). 

If you just want to use the scatnet representation for classification, you may want to use the [`stack_scat_output`](https://github.com/oinegue/scatnetgpu/blob/master/scatnetgpu/scatnetgpu.py#L508) method that returns all signals and all paths. Just [`.flatten`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.flatten.html) the returned signals and you will have a features vector.

**Comment by deepakanandece** [comment]
Hello,
 I would like to use the scanet output as multi-channel image, rather than flatenning. Flatenning will distroy the spatial relation. Can you suggest the way in which i get output as a scaled and thick(multi-channel) image?

**Comment by oinegue** [comment]
Just use [`stack_scat_output`](https://github.com/oinegue/scatnetgpu/blob/master/scatnetgpu/scatnetgpu.py#L508)`(sn.transform(img))
`. It returns a tuple where the first element is what you are looking for. Please look at the source code to understand the insights.
